lib/ecto/query/planner.ex

```diff
@@ -646,6 +664,16 @@ defmodule Ecto.Query.Planner do
   defp expr_to_cache(%QueryExpr{expr: expr}), do: expr
   defp expr_to_cache(%SelectExpr{expr: expr}), do: expr
 
+  defp cast_and_merge_params(:where, _query, %BooleanExpr{expr: {:in, [], [_left, %Ecto.SubQuery{params: p}]}}, params, _adapter) do
```

Comment:
Similarly, this only works if the `:in` is at the root. We also need to return the subquery cache key.